org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/impl/AbstractFSSynchronizationTest.java

```diff
@@ -0,0 +1,427 @@
+import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.xtext.generator.OutputConfiguration;
+import org.eclipse.xtext.ui.XtextProjectHelper;
+import org.eclipse.xtext.xbase.lib.ArrayExtensions;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author kosyakov - Initial contribution and API
+ */
+public abstract class AbstractFSSynchronizationTest extends AbstractBuilderParticipantTest {
+	private static String PROJECT_NAME = "myProject";
+
+	private IProject project;
+
+	@Before
+	public void createJavaProject() throws Exception {
+		project = createJavaProject(AbstractFSSynchronizationTest.PROJECT_NAME).getProject();
+		addNature(project, XtextProjectHelper.NATURE_ID);
+	}
+
+	@After
+	public void forgetProject() throws Exception {
+		project = null;
+	}
+
+	@Test
+	public void testUpdateFileContent() {
+		testUpdateFileContent(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testUpdateFileContentWithProjectAsOutput() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testUpdateFileContent(project));
+	}
+
+	protected void testUpdateFileContent(IContainer output) {
+		createJavaIoFile(output.getFile(new Path("Foo.txt")).getLocation(), "object Bar");
+		createFile(project.getFile(("src/Foo" + F_EXT)).getFullPath(), "object Foo");
+		build();
+		Assert.assertEquals("object Foo", readFile(output.getFile(new Path("Foo.txt"))));
+	}
+
+	@Test
+	public void testTouchFile() {
+		testTouchFile(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testTouchFileWithProjectAsOutput() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testTouchFile(project));
+	}
+
+	protected void testTouchFile(IContainer output) {
+		createJavaIoFile(output.getFile(new Path("Foo.txt")).getLocation(), "object Foo");
+		createFile(project.getFile(("src/Foo" + F_EXT)).getFullPath(), "object Foo");
+		build();
+		Assert.assertEquals("object Foo", readFile(output.getFile(new Path("Foo.txt"))));
+	}
+
+	@Test
+	public void testCreateFile() {
+		testCreateFile(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testCreateFileWithProjectAsOutput() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testCreateFile(project));
+	}
+
+	protected void testCreateFile(IContainer output) {
+		try {
+			IFile file = createFile(project.getFile(("src/Foo" + F_EXT)).getFullPath(), "object Foo");
+			build();
+			Assert.assertEquals("object Foo", readFile(output.getFile(new Path("Foo.txt"))));
+			File javaIoFile = output.getFile(new Path("Foo.txt")).getLocation().toFile();
+			javaIoFile.delete();
+			javaIoFile.getParentFile().delete();
+			file.touch(monitor());
+			build();
+			Assert.assertEquals("object Foo", readFile(output.getFile(new Path("Foo.txt"))));
+		} catch (CoreException e) {
+			throw Exceptions.sneakyThrow(e);
```

Comment:
In tests it would be safe to throw new RuntimeException or Assert.fail.
But it's only test code, so I'm fine with this.